api/http/http.go

```diff
@@ -1,14 +1,71 @@
-// Copyright 2014 Canonical Ltd.
+// Copyright 2022 Canonical Ltd.
 // Licensed under the AGPLv3, see LICENCE file for details.
 
 package http
 
 import (
+	"context"
+	"io"
 	"net/http"
+	"net/url"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/api/base"
 )
 
 // HTTPClient is an API-specific HTTP client.
 type HTTPClient interface {
 	// Do sends the HTTP request, returning the subsequent response.
 	Do(req *http.Request) (*http.Response, error)
 }
+
+// HTTPDoer exposes the functionality of httprequest.Client needed here.
+type HTTPDoer interface {
+	// Do sends the given request.
+	Do(context context.Context, req *http.Request, resp interface{}) error
+}
+
+// OpenURIFunc returns a reader for the blob at the specified uri.
+type OpenURIFunc func(uri string, query url.Values) (io.ReadCloser, error)
+
+// NewURIOpener returns a URI opener func for the api caller.
+func NewURIOpener(apiCaller base.APICaller) OpenURIFunc {
+	return func(uri string, query url.Values) (io.ReadCloser, error) {
+		return OpenURI(apiCaller, uri, query)
+	}
+}
+
+// OpenURI performs a GET on a Juju HTTP endpoint returning the specified blob.
+func OpenURI(apiCaller base.APICaller, uri string, query url.Values) (io.ReadCloser, error) {
```

Comment:
I don't like this function signature.

This should take a context and a HTTPDoer not an apiCaller. It's a heavy object that makes testing the underlying premise hard and our reliance on gomock more.

Reply:
done